### ISSUE TITLE
proper support of cancellation in resolveCompletionItem

### ISSUE THREAD

**Issue by heejaechang** [microsoft/vscode#95485]
- VSCode Version: 1.44.0
- OS Version: 1909

Steps to Reproduce:

1. install python 3.7+
2. python -m pip install pandas
3. clone https://github.com/microsoft/pyright
4. open vscode with pyright repo root
5. run "pyright client"
6. type "import pandas as pd"
7. in next line, type "pd.read_c" 
8. wait tooltip for read_csv from completion list
9. while it is calculating (resolveCompletionItem call), move to other completion item causing resolveCompletionItem call to be cancelled.

Does this issue occur when all extensions are disabled?: No

..

issue seems with this code
https://github.com/microsoft/vscode/blob/master/src/vs/workbench/api/common/extHostLanguageFeatures.ts#L904

this resolveCompletionItem call seems to not handle cancellation properly. so even if cancellation happened, it just take whatever result it got and cache it causing until next (new) completion list, tooltip is lost for one that got cancelled.

basically, it looks like "then" should handle error case as well.

**Comment by jrieken** [comment]
Not sure what you mean? The actual caching of the resolve-result happens in the UX, here https://github.com/microsoft/vscode/blob/568466d2fe7152698dc9b094ced20b5cbe7693fa/src/vs/editor/contrib/suggest/suggest.ts#L97 and cancellation is handled. Is it that your language service throws an error (which is not the [cancellation error](https://github.com/microsoft/vscode/blob/568466d2fe7152698dc9b094ced20b5cbe7693fa/src/vs/editor/contrib/suggest/suggest.ts#L103)) which is then cached?

**Comment by jrieken** [comment]
fyi @dbaeumer

**Comment by heejaechang** [comment]
hmmm.. code seems changed. it now uses await rather than .then(...)

the code I shared used ".then" without error handling part. with await, I guess as long as resolveCompletionItem promise reject, things will work as expected?

https://github.com/microsoft/vscode/commit/8f1b2e565797929d8e8fd772ed35fa202ac76567#diff-bf62e0eff390ee2bc5596fbc5b6bf2e7L904

**Comment by jrieken** [comment]
Using await or then shouldn't matter here as no error-handling exists here. Everything should happen [here](https://github.com/microsoft/vscode/blob/568466d2fe7152698dc9b094ced20b5cbe7693fa/src/vs/editor/contrib/suggest/suggest.ts#L106). There we cache the request-promise (which is the whole thing, request to ext host, request to extension, etc) and we [clear](https://github.com/microsoft/vscode/blob/568466d2fe7152698dc9b094ced20b5cbe7693fa/src/vs/editor/contrib/suggest/suggest.ts#L109) the cached promise when cancellation happens, unless resolving is already done

**Comment by heejaechang** [comment]
I see. probably due to "https://github.com/microsoft/vscode-languageserver-node/issues/615" not returning error but return item.

it just got fixed, so that might be the root cause then.

**Comment by jrieken** [comment]
ok - closing as duplicate of the LSP issue
